Update security manager section (#951)
MicroEJ/docs: KernelDeveloperGuide/securityManagement.rst

Security Management
===================

The security management can be done in various ways, the most common way of declaring a security management policy is by the usage of the `SecurityManager`_ class.

> For the sake of ROM footprint optimization, calls to Permission checks are disabled by default.

Reviewer: Not clear: call to permission check are not defined (though it is difficult to understand why they are disabled).

This can be achieved by subclassing the base `SecurityManager`_ class, overriding its `SecurityManager.checkPermission(Permission)`_ method,
and registering an instance of this class to the Kernel by a call to `System.setSecurityManager(SecurityManager)`_.

.. code-block:: java

      // create a new Security Manager
      SecurityManager sm = new SecurityManager() {
         @Override
         public void checkPermission(java.security.Permission perm) {
            // here implement your Kernel Security Policy
         };
      };
      // register the Security Manager
      System.setSecurityManager(sm);

Then you have to implement your own Security Policy.

Implementation examples
-----------------------

MicroEJ provides several security manager implementation examples in the `com.microej.library.util.kf-util`_ (2.8.0+) module.
A usage example of these implementations is available in the `Kernel-GREEN`_ GitHub project.

Kernel Security Policy Manager
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This implementation is inspired from the `JavaPolicyFile`_ concept.
Features come with a policy file that describes the permission they will need at runtime.

> This concept assumes that the system owner is able to verify the content of this file before the feature is deployed onto the Kernel (using forge connect for instance).

Reviewer: Forge Connect is not defined in the documention

> Security Management

Reviewer 2: It remains:
```
The security management can be done in various ways, [...]
```
that I would change into
```
Enforcing an Application Security Policy can be done in various ways, [...]
```

> Security Management

Reviewer 2: Fixed

===================

> The security management can be done in various ways, the most common way of declaring a security management policy is by the usage of the `SecurityManager`_ class.

Reviewer 2: ok..

For the sake of ROM footprint optimization, permission checks calls are disabled by default to avoid extra code processing if the system owner does not want to use the Security Manager.

> In order to activate this feature the :ref:`option_enable_security_manager` option must be set.

Author: I think its interesting from the Kernel developer POV to know why its disabled by default, and as this is located in KernelDeveloperGuide, I'm pretty sure this is where it should be said (app developer doesn't need really to know about this) a section explaining how to setup a security check in a foundation is already available here: https://docs.microej.com/en/latest/KernelDeveloperGuide/kfEnabledLibraries.html?highlight=permission#implement-a-security-manager-permission-check
Unless I'm misunderstanding something here ?

> In order to activate this feature the :ref:`option_enable_security_manager` option must be set.

Reviewer 2: I would think it is only useful to users that develop Kernel access-protected libraries.
And it is not useful to users that develop Kernel security policy (i.e. implement a security manager).
But it is fine.

Once the security manager checks are enabled, you can then implement your own policy.

This can be achieved by subclassing the base `SecurityManager`_ abstract class, overriding its `SecurityManager.checkPermission(Permission)`_ method,

> and registering an instance of this class to the Kernel by a call to `System.setSecurityManager(SecurityManager)`_.

Reviewer 2: Suggestion applied, thanks!

.. code-block:: java

>       // create a new Security Manager

Reviewer 2: Suggestion applied, thanks!

      SecurityManager sm = new SecurityManager() {
         @Override
         public void checkPermission(java.security.Permission perm) {

>             // here implement your Kernel Security Policy

Reviewer 2: Suggestion applied, thanks!

         };
      };

>       // register the Security Manager

Reviewer 2: Suggestion applied, thanks!

      System.setSecurityManager(sm);

> Then you have to implement your own Security Policy.

Reviewer 2: Suggestion applied, thanks!

Implementation examples
-----------------------

> MicroEJ provides several security manager implementation examples in the `com.microej.library.util.kf-util`_ (2.8.0+) module.

Reviewer 2: Yes sure, you can adapt my suggestion for 2 instances.
I was initially confused by the naming of those security policy managers.

> A usage example of these implementations is available in the `Kernel-GREEN`_ GitHub project.

Reviewer 2: Suggestion applied, thanks!

> Kernel Security Policy Manager

Author: the module has indeed already been released but yeah, it is meant to monitor apps permissions

> Kernel Security Policy Manager

Reviewer 2: Ok :/

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> This implementation is inspired from the `JavaPolicyFile`_ concept.

Reviewer 2: Suggestion applied, thanks!

> Features come with a policy file that describes the permission they will need at runtime.

Reviewer 2: Suggestion applied, thanks!

> By default the name of this file should be ``feature.policy.json`` you can override the default name using the property ``feature.policy.name`` to be added in the Kernel `properties`_.

Reviewer 2: Suggestion applied, thanks!

This concept assumes that the system owner is able to verify the content of this file before the feature is deployed onto the Kernel (using an application store system for instance).

> The file is then parsed on the feature installation and the permissions are stored in a map in the security manager.

Reviewer 2: Suggestion applied, thanks!

> Finally, when a permission check is requested, the security manager will check if the feature has specified the requested permission or not in which case it throws a `SecurityException`_.

Reviewer 2: The KF Security Policy Manager is responsible for checking the permission has been granted to the calling Application on permission checks requested by the system libraries.

Reviewer 2: Suggestion applied, thanks!

          "<permissionName2>":["<permissionAction1>"]
        },
        "<permissionClassName2>":{
          "<permissionName3>":["<permissionAction3>"]
        }
      }
    }

Here is now an example of what a real JSON file can look like:

.. code block:: JSON
    {
      "permissions": {
        "ej.microui.display.DisplayPermission":{
          "*":[]
        },
        "ej.microui.event.EventPermission":{
          "null":["null"]
        },
        "ej.microui.display.FontPermission":{},
        "ej.microui.display.ImagePermission":{
          "null":["*"]
        },"ej.microui.MicroUIPermission":{
          "*":["start"]
        },"java.net.SocketPermission":{
          "www.microej.com":["connect","resolve"]
        },"java.util.PropertyPermission":{
          "property":["write","read"]
        },"java.lang.RuntimePermission":{
          "exit":[]
        }
      }
    }

.. note::
    The permission ``name`` attribute is specific to the permission implementation therefore, each permission has its own definition of what a name is.

The parser contains two key words to allow more flexibility over the content of the file:

> - "*": the wildcard symbol represents ``everything`` it can be used for permission class name, permission name and permission actions.

Reviewer 2: ```suggestion
- "*": the wildcard symbol means "any". It can be used for permission class name, permission name and permission actions.
```

Reviewer 2: Suggestion applied, thanks!

> - "null": the ``null`` keyword represents a java ``null`` value, it can be used for permission name and permission actions.

Reviewer 2: ```suggestion
- "null": the ``null`` keyword represents a Java ``null`` value. It can be used for permission name and permission actions.
```

Reviewer 2: Suggestion applied, thanks!

To simplify the file structure you can also choose to have an empty object value for permission className or/and permission actions such as shown in the example above:

.. code block:: JSON
    {
      "permissions": {
        "ej.microui.display.DisplayPermission":{
          "*":[]
        },
        "ej.microui.display.FontPermission":{},
        "java.lang.RuntimePermission":{
          "exit":[]
          }
        }
    }

This example:
- allows the usage of any permission name and any actions for the ``ej.microui.display.DisplayPermission`` permission.
- allows the usage of any permission name and any actions for the ``ej.microui.display.FontPermission`` permission.
- allows the ``exit`` permission name and any actions for the ``java.lang.RuntimePermission`` permission.

> Describing ``anything`` either with an empty value or a ``*`` is left to the developer preference and should be processed in the exact same way by the security manager.

Reviewer 2: ```suggestion
Using an empty value or the ``*`` wildcard is left to the developer preference and should be processed in the exact same way by the security policy resource loader.
```

Reviewer 2: Fixed

.. warning::

>     Make sure that specified permission class names are embedded by declaring them as `requiredTypes`.
>     Any permission check done on a permission class without embedded name will result in a `SecurityException`_.

Reviewer 2: ```suggestion
    If the Kernel does not embed all class names (see :ref:`stripping-class-names-from-an-application`),
    make sure that specified permission class names are embedded by declaring them as :ref:`Required Types <types>`.
    Any permission check done on a permission class without embedded name will result in a `SecurityException`_.
```
Note: shouldn't this result in a error (or at least warning) at installation time (when loading the policy) ?

Author: The check is not actually done on class type to allow more flexiblity in case the permission classes should be added later on in the system, 
We're using getClass.getName(), to check for the equality of the requested access / authorized ones, it is not actually an issue because in case the name is not embedded, the equality will just fail and throws the Security Exception and be handled the same way.

Reviewer 2: > The check is not actually done on class type to allow more flexiblity in case the permission classes should be added later on in the system,

Do you mean at installation time ?

If so, I don't see the use case: permission checks should only concern the libraries included in the Kernel, right ?
We don't have dynamic class loading there.

Reviewer 2: But ok, I guess this should be addressed in the KF-Util library.
Here the documentation complies with the current implementation.

Author: Yeah I see your point, with some hindsight I'm not sure this is actually a coherent usecase but we designed it to be as flexible as possible

Kernel Security Manager
~~~~~~~~~~~~~~~~~~~~~~~

> This implementation is based on the creation of `FeaturePermissionCheckDelegate`_ to specify the behavior of the `SecurityManager.checkPermission(Permission)`_ for each permission.

Reviewer 2: ```suggestion
This implementation is based on the creation of `FeaturePermissionCheckDelegate`_ to specify the behavior of the `SecurityManager.checkPermission(Permission)`_ for each permission class.
```

Reviewer 2: Fixed

> If a permission check is done and no delegate for this permission exist, a `SecurityException`_ is thrown.

Reviewer 2: ```suggestion
If a permission check is done and no delegate for its permission class is found, a `SecurityException`_ is thrown.
```

Reviewer 2: Fixed

> An usage example of this implementation and further explanations about how to implement your own delegates are demonstrated in the `Kernel-GREEN`_ project.
> This Kernel implements a logging-only security policy by default.

Reviewer 2: ```suggestion
An example of this policy manager is used by the `Kernel-GREEN`_. The policy grants all applications the permission for a list of permission classes and logs all protected accesses by Applications.
```

Reviewer 2: Fixed
